r/R/dplyr-join.R

```diff
@@ -83,7 +94,17 @@ full_join.arrow_dplyr_query <- function(x,
                                         suffix = c(".x", ".y"),
                                         ...,
                                         keep = FALSE) {
-  do_join(x, y, by, copy, suffix, ..., keep = keep, join_type = "FULL_OUTER")
+  query <- do_join(x, y, by, copy, suffix, ..., keep = keep, join_type = "FULL_OUTER")
+
+  # If we are doing a full outer join and not keeping the join keys of
+  # both sides, we need to coalesce. Otherwise, rows that exist in the
+  # RHS will have NAs for the join keys.
+  if (!keep) {
+    query$selected_columns <- post_join_projection(names(x), names(y), handle_join_by(by, x, y), suffix)
+    collapse.arrow_dplyr_query(query)
```

Comment:
I don't think you need this collapse() here since do_join() collapses for you and you're just projecting after. (Moreover, since you aren't assigning the result here and return `query` below, the result of this collapse isn't being used.)

```suggestion
```